[tune] Adds option to checkpoint at end of trials

praveen-palanisamy opened the pull request:
## What do these changes do?

Adds a Tune experiment option to enable checkpointing at the end of trials in-addition-to/irrespective of the usual checkpointing done periodically at checkpoint_freq. The option configuration is consistent with the Experiment interface and the config_parser through which the `checkpoint_freq` is set currently.


Example usage:

```python
run_experiments({
    "my_experiment_name": {
            "run": my_trainable
            "checkpoint_freq": 10,
            "checkpoint_at_end": True,
            "max_failures": 5,
        },
 })
```
## Related issue number

This PR addresses issue #2740, #2742

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7800/
Test PASSed.

richardliaw commented on python/ray/tune/trial_runner.py:
```diff
@@ -234,13 +235,16 @@ def _process_events(self):
 
             if decision == TrialScheduler.CONTINUE:
-                if trial.should_checkpoint():
+                if trial.should_checkpoint(result):
```
We can actually just take this out of the if-else right? 
```
if trial.should_checkpoint(result):
    self.trial_executor.save(trial)

if decision == TrialScheduler.CONTINUE:
    self.trial_executor.continue_training(trial)
elif decision == ...
```

praveen-palanisamy replied:
Yeah! That makes it cleaner. I have pushed the changed. I left the comment with the TODO (blocking call to save) for you untouched :D

richardliaw commented on python/ray/tune/trial_runner.py:
```diff
@@ -234,13 +235,16 @@ def _process_events(self):
             elif decision == TrialScheduler.PAUSE:
                 self.trial_executor.pause_trial(trial)
             elif decision == TrialScheduler.STOP:
+                # Checkpoint the state before ending the trial if checkpoint_at_end experiment option is set to True
```
you may want to run `flake8 --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504` on this file

richardliaw commented on python/ray/tune/experiment.py:
```diff
@@ -43,6 +43,8 @@ class Experiment(object):
         checkpoint_freq (int): How many training iterations between
             checkpoints. A value of 0 (default) disables checkpointing.
+        checkpoint_at_end (bool): Whether to checkpoint at the end of the
+            experiment. Default is False.
```
Can you also note that this is only enabled if checkpoint_freq is not 0?

ericl replied:
You mean is enabled regardless right?

praveen-palanisamy replied:
Currently, it will be disabled if checkpoint_freq is 0 as Richard pointed out. Do we want to override and enable this even when checkpoint_freq is 0?

praveen-palanisamy replied:
I think it will be better to have the checkpoint_at_end option have an effect even if checkpoint_freq is set to 0/default. This will provide more flexibility for the users depending on the use cases. What do you guys think?

richardliaw replied:
The new change looks good!

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7815/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7813/
Test PASSed.

richardliaw commented on python/ray/tune/test/trial_runner_test.py:
```diff
@@ -938,6 +938,29 @@ def testCheckpointing(self):
+        runner.step()
+        self.assertEqual(trials[0].last_result[DONE], True)
+        path = runner.trial_executor.save(trials[0])
+        #self.assertEqual(path[-1], 3)
```
Can you remove this comment?

praveen-palanisamy replied:
Sure. Done.

richardliaw commented on python/ray/tune/trial_runner.py:
```diff
@@ -233,10 +234,11 @@ def _process_events(self):
-                if trial.should_checkpoint():
+            if trial.should_checkpoint(result):
                     # TODO(rliaw): This is a blocking call
                     self.trial_executor.save(trial)
```
I think this line should be decremented?

praveen-palanisamy replied:
Oops :/. Fixed the indentation. Thanks.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7826/
Test FAILed.

richardliaw commented:
Looks like the new checkpoint test is failing because the trainable used in
the test did not implement _save
On Tue, Aug 28, 2018 at 8:06 AM UCB AMPLab <notifications@github.com> wrote:

> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7826/
> Test FAILed.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/2754#issuecomment-416621864>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEUc5TYTIadLNjdSvG9zFl7jb-d49he5ks5uVVxngaJpZM4WOugL>
> .
>

praveen-palanisamy commented:
Hmmm. 
Will not such trainable (without _save implemented) cause that issue for checkpointing in general?

 I wonder if the testFailureRecovery* cases are also failing due to this? Or are they actually asserting the failures due to mock error and not affected by this at all?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7827/
Test FAILed.

praveen-palanisamy commented:
To handle it, I'm thinking of something like this in train_runner:

```python
if trial.should_checkpoint(result):
    try:
        # TODO(rliaw): This is a blocking call
        self.trial_executor.save(trial)
    except NotImplementedError:
        print("Trainable does not have _save implemented."
              "No checkpoints will be saved.")
```
What do you think?

richardliaw commented:
Actually, on a second thought; I'm not sure if that will work, because at least for Ray, `_save` is implemented on the Actor and therefore will raise RayGetError rather than a NotImplementedError...

praveen-palanisamy commented:
But the remote function is still the trainable right? Which raises NotImplementedError in the first place?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7828/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7832/
Test FAILed.

richardliaw commented:
Using the MockAgent works for now.  The proper fix for that would probably have RayGetError dynamically subclass the thrown exception, but we can deal with that in a future issue. In any case, can you revert the try-block added previously?

Looks like errors are failing due to `bool("False") == True`, resulting in checkpointing always happening (see https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse).

One option here to fix this would be 
1. to turn the added argument into a flag, as suggested in the post above, and
2. Modify `to_argv` in `config_parser.py`:

```python
def to_argv(config):
    """Converts configuration to a command line argument format."""
    argv = []
    for k, v in config.items():
        if "-" in k:
            raise ValueError("Use '_' instead of '-' in `{}`".format(k))
        argv.append("--{}".format(k.replace("_", "-")))
        if isinstance(v, string_types):
            argv.append(v)
        elif isinstance(v, bool):
            pass  # Treat booleans as flags
        else:
            argv.append(json.dumps(v, cls=_SafeFallbackEncoder))
    return argv
```

praveen-palanisamy commented:
I reverted the try-except block.
I also reverted the test case modification and simplified the checks and tested it.
The below test code works as expected:

### 1. Test when `checkpoint_at_end is True` 
(The last line prints True)
```python
from ray.tune.suggest import BasicVariantGenerator
from ray.tune.trial import Trial

runner = TrialRunner(BasicVariantGenerator())
kwargs = {
    "stopping_criterion": {
        "training_iteration": 2
    },
    "checkpoint_freq": 5,
    "checkpoint_at_end": True
}
runner.add_trial(Trial("__fake", **kwargs))
trials = runner.get_trials()

runner.step()
runner.step()
runner.step()
print("Status:", trials[0].status)

print("trial_has_checkpoint?:", trials[0].has_checkpoint())
```
### 2. Test when `checkpoint_at_end is False` 
(The last line prints False)
```python
from ray.tune.suggest import BasicVariantGenerator
from ray.tune.trial import Trial

runner = TrialRunner(BasicVariantGenerator())
kwargs = {
    "stopping_criterion": {
        "training_iteration": 2
    },
    "checkpoint_freq": 5,
    "checkpoint_at_end": False
}
runner.add_trial(Trial("__fake", **kwargs))
trials = runner.get_trials()

runner.step()
runner.step()
runner.step()
print("Status:", trials[0].status)

print("trial_has_checkpoint?:", trials[0].has_checkpoint())
```

praveen-palanisamy commented:
@richardliaw : Where are you seeing the errors/fails due to  `bool("False") == True` ? 

I do not seem to encounter that situation anywhere. Both the JSON's bool (`true`) and python's bool (`True`) work well for the option this PR introduces right? I don't think we use strings (`"True"`) anywhere or am I missing something?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7837/
Test FAILed.

richardliaw commented:
Yeah - try running `python examples/hyperopt_example.py --smoke-test` which should pass on master but currently errors on this branch.

Resolving this last issue should allow Jenkins to pass (Jenkins console output here: https://amplab.cs.berkeley.edu/jenkins/job/Ray-PRB/7837/console)

Let me know if you have any questions!

praveen-palanisamy commented:
Thanks for the pointer to the example. That helped to see what you meant. I went with another option to fix it. By just skipping the the conversion to string (`json.dumps`). Looks fine?

praveen-palanisamy commented:
I reverted 7830051 because that seems to pass the previously failing tests locally. 
Both 
- ` python -m pytest ray/tune/test/trial_runner_test.py` and 
- ` python ray/rllib/train.py --env CartPole-v0 --run PG --stop '{"training_iteration": 2}' --config '{"sample_batch_size": 500, "num_workers": 1}'`
passed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7843/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7845/
Test PASSed.

praveen-palanisamy commented:
@richardliaw : I am not sure why the Travis build is failing. Didn't notice any errors in the log. Is there something more that needs to be done for this PR?

richardliaw commented:
Got it. 

RE: tests - there was a linting error that I just fixed.
RE: Reversion of redundant checkpoint - we can leave it as is, but I agree. The trial runner is a bit cluttered, and we should probably address that in the context of #2561 (if you want to give it a shot, that'd be awesome, but don't feel obligated to do it).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7861/
Test PASSed.

richardliaw commented:
Looks good; test errors seem unrelated. Thanks for contributing this @praveen-palanisamy!
